app/src/main/java/fly/speedmeter/grub/MainActivity.java

```diff
@@ -74,13 +74,44 @@ protected void onCreate(Bundle savedInstanceState) {
         onGpsServiceUpdate = new Data.onGpsServiceUpdate() {
             @Override
             public void update() {
-                maxSpeed.setText(data.getMaxSpeed());
-                distance.setText(data.getDistance());
+                double maxSpeedTemp = data.getMaxSpeed();
+                double distanceTemp = data.getDistance();
+                double averageTemp;
                 if (sharedPreferences.getBoolean("auto_average", false)){
-                    averageSpeed.setText(data.getAverageSpeedMotion());
+                    averageTemp = data.getAverageSpeedMotion();
                 }else{
-                    averageSpeed.setText(data.getAverageSpeed());
+                    averageTemp = data.getAverageSpeed();
                 }
+
+                String speedUnits;
+                String distanceUnits;
+                if (sharedPreferences.getBoolean("miles_per_hour", false)) {
+                    maxSpeedTemp *= 0.62137119;
+                    distanceTemp = distanceTemp / 1000.0 * 0.62137119;
+                    averageTemp *= 0.62137119;
+                    speedUnits = "mi/h";
+                    distanceUnits = "mi";
+                } else {
+                    speedUnits = "km/h";
+                    if (distanceTemp <= 1000) {
+                        distanceTemp /= 1000.0;
```

Comment:
the distance is in meter, it mean that you must divide by 1000 when you are in km.